internal/service/wallet.go

```diff
@@ -226,37 +223,122 @@ func (s *walletService) GetWalletTransactions(ctx context.Context, walletID stri
+
+	var idempotencyKey string
+	if lo.FromPtr(req.IdempotencyKey) != "" {
+		idempotencyKey = lo.FromPtr(req.IdempotencyKey)
+	} else {
+		idempotencyKey = types.GenerateUUID()
+	}
+
+	// Prepare credit operation details
+	referenceType := types.WalletTxReferenceTypeExternal
+	referenceID := idempotencyKey
+
+	// Handle special case for purchased credits with invoice
+	if req.TransactionReason == types.TransactionReasonPurchasedCreditInvoiced {
+		paymentID, err := s.handlePurchasedCreditInvoicedTransaction(
+			ctx,
+			walletID,
+			lo.ToPtr(idempotencyKey),
+			req,
+		)
+		if err != nil {
+			return nil, err
+		}
+		referenceID = paymentID
+	}
+
+	// Create wallet credit operation
 	creditReq := &wallet.WalletOperation{
 		WalletID:          walletID,
 		Type:              types.TransactionTypeCredit,
-		CreditAmount:      req.Amount,
+		CreditAmount:      req.CreditsToAdd,
 		Description:       req.Description,
 		Metadata:          req.Metadata,
-		TransactionReason: types.TransactionReasonFreeCredit,
-		ReferenceType:     types.WalletTxReferenceTypeRequest,
-		ReferenceID:       types.GenerateUUIDWithPrefix(types.UUID_PREFIX_WALLET_TRANSACTION),
+		TransactionReason: req.TransactionReason,
+		ReferenceType:     referenceType,
+		ReferenceID:       referenceID,
 		ExpiryDate:        req.ExpiryDate,
+		IdempotencyKey:    idempotencyKey,
 	}
 
-	if req.PurchasedCredits {
-		if req.GenerateInvoice {
-			creditReq.TransactionReason = types.TransactionReasonPurchasedCreditInvoiced
-		} else {
-			creditReq.TransactionReason = types.TransactionReasonPurchasedCreditDirect
-		}
+	// Process wallet credit
+	if err := s.CreditWallet(ctx, creditReq); err != nil {
+		return nil, err
 	}
 
-	if creditReq.ReferenceID == "" || creditReq.ReferenceType == "" {
-		creditReq.ReferenceID = types.GenerateUUIDWithPrefix(types.UUID_PREFIX_WALLET_TRANSACTION)
-		creditReq.ReferenceType = types.WalletTxReferenceTypeRequest
+	return s.GetWalletByID(ctx, walletID)
+}
+
+func (s *walletService) handlePurchasedCreditInvoicedTransaction(ctx context.Context, walletID string, idempotencyKey *string, req *dto.TopUpWalletRequest) (string, error) {
+	// Initialize required services
+	invoiceService := NewInvoiceService(s.ServiceParams)
+	customerService := NewCustomerService(s.ServiceParams)
+	paymentService := NewPaymentService(s.ServiceParams)
+
+	// Retrieve wallet and customer details
+	w, err := s.WalletRepo.GetWalletByID(ctx, walletID)
+	if err != nil {
+		return "", err
 	}
 
-	if err := s.CreditWallet(ctx, creditReq); err != nil {
-		return nil, err
+	customer, err := customerService.GetCustomer(ctx, w.CustomerID)
+	if err != nil {
+		return "", err
 	}
 
-	return s.GetWalletByID(ctx, walletID)
+	var paymentID string
+	err = s.DB.WithTx(ctx, func(ctx context.Context) error {
+		// Create invoice for credit purchase
+		invoice, err := invoiceService.CreateInvoice(ctx, dto.CreateInvoiceRequest{
+			CustomerID:     customer.ID,
+			AmountDue:      req.CreditsToAdd,
```

Comment:
Need to convert the Credits Amount back to Currency amount
`amount in the currency =  number of credits * conversion_rate`